README.md
# Sunrise Sunset

Dart API that provides sunset and sunrise times for a given latitude and longitude.  This API is a wrapper around the 
free [Sunset and Sunrise Times API](https://sunrise-sunset.org/)

![Phases of the sun and moon](https://sunrise-sunset.org/graph.svg)

## Getting started

You should ensure that you add the dependency in your project.
```yaml
dependencies:
 sunrise_sunset: "^1.0.2"
```

You should then run `flutter packages upgrade` or update your packages in IntelliJ.

## Usage

Import the file:

```
import 'package:sunrise_sunset/sunrise_sunset.dart';
```

To use simply call the static method `SunriseSunset.getResults(...)`.

Time values in response will be expressed `Datetime` instances and `day_length` will be expressed in seconds.

```
final response = await SunriseSunset.getResults(date: DateTime.now(), latitude: 36.7201600, longitude: -4.4203400);
```

### Parameters

Here is a list of supported parameters to use in the `SunriseSunset.getResults()` method

 | Name          | Optional                                                | 
 |---------------|---------------------------------------------------------|
 | latitude      | No                                                      | 
 | longitude     | No                                                      | 
 | date          | Yes (If not present, date defaults to current date)     | 








CHANGELOG.md
## [1.0.0]

* First release.

## [1.0.1]

* Added example 
* Simplified API

## [1.0.2]

* Updated CHANGELOG.md
* Enhanced the README.md 
* Added helpful graphic
* Gave attribution to source REST API

## [1.0.3]

* Updated README.md properties table

## [1.0.4]

* Bug with Date param fixed

test/sunrise_sunset_test.dart
import 'package:test/test.dart';

import 'package:sunrise_sunset/sunrise_sunset.dart';

void main() {
  test(
    'Test OK response without format',
    () async {
      final response = await SunriseSunset.getResults(
          latitude: 39.001735, longitude: -119.752744);
      expect(response.success, true);
      expect(response.error, isEmpty);
      expect(response.data, isNotNull);
      expect(response.data.dayLength, 47525);
      expect(response.data.astronomicalTwilightEnd,
          DateTime.parse("2020-04-14T04:10:06+00:00"));
    },
  );

  test(
    'Test INVALID_REQUEST',
    () async {
      final response = await SunriseSunset.getResults(longitude: 150);
      expect(response.success, false);
      expect(response.error, isNotEmpty);
      expect(response.data, isNull);
      expect(response.error.contains("lat"), true);
    },
  );

  test('Test all values', () async {
    var response = await SunriseSunset.getResults(
        date: DateTime.now().toLocal(),
        latitude: 39.001735,
        longitude: -119.752744);
    print(response);
  });
}


test/mock/mock_data_service.dart
import 'package:sunrise_sunset/src/service/data_service.dart';
import 'package:sunrise_sunset/src/service/options.dart';

class MockDataService implements DataRepository {
  Map data = mock0;

  @override
  Future<SunriseSunsetResponse<T>> serverRequest<T>(Options options) async {
    return parseData(data, options);
  }
}

//OK formatted results
Map mock0 = {
  "results": {
    "sunrise": "7:27:02 AM",
    "sunset": "5:05:55 PM",
    "solar_noon": "12:16:28 PM",
    "day_length": "9:38:53",
    "civil_twilight_begin": "6:58:14 AM",
    "civil_twilight_end": "5:34:43 PM",
    "nautical_twilight_begin": "6:25:47 AM",
    "nautical_twilight_end": "6:07:10 PM",
    "astronomical_twilight_begin": "5:54:14 AM",
    "astronomical_twilight_end": "6:38:43 PM"
  },
  "status": "OK"
};

//OK no formatted results
Map mock1 = {
  "results": {
    "sunrise": "2015-05-21T05:05:35+00:00",
    "sunset": "2015-05-21T19:22:59+00:00",
    "solar_noon": "2015-05-21T12:14:17+00:00",
    "day_length": 51444,
    "civil_twilight_begin": "2015-05-21T04:36:17+00:00",
    "civil_twilight_end": "2015-05-21T19:52:17+00:00",
    "nautical_twilight_begin": "2015-05-21T04:00:13+00:00",
    "nautical_twilight_end": "2015-05-21T20:28:21+00:00",
    "astronomical_twilight_begin": "2015-05-21T03:20:49+00:00",
    "astronomical_twilight_end": "2015-05-21T21:07:45+00:00"
  },
  "status": "OK"
};

//INVALID_REQUEST
Map mock2 = {"status": "INVALID_REQUEST"};

//INVALID_DATE
Map mock3 = {"status": "INVALID_DATE"};

//UNKNOWN_ERROR
Map mock4 = {"status": "UNKNOWN_ERROR"};


example/lib/main.dart
import 'package:sunrise_sunset/sunrise_sunset.dart';

void main() async {
  try {
    final response = await SunriseSunset.getResults(
        latitude: 39.001735, longitude: -119.752744);

    if (response.success) {
      print('Sunrise: ${response.data.sunrise}');
      print('Sunset: ${response.data.sunset}');
      print('Solar Noon: ${response.data.solarNoon}');
      print('Day Length: ${response.data.dayLength}');
      print('Civil Twilight Start: ${response.data.civilTwilightBegin}');
      print('Civil Twilight End: ${response.data.civilTwilightEnd}');
      print('Nautical Twilight Start: ${response.data.nauticalTwilightBegin}');
      print('Nautical Twilight End: ${response.data.nauticalTwilightEnd}');
      print(
          'Astronomical Twilight Start: ${response.data.astronomicalTwilightBegin}');
      print(
          'Astronomical Twilight End: ${response.data.astronomicalTwilightEnd}');
    } else {
      print(response.error);
    }
  } catch (err) {
    print("Failed to get data.");
    print(err);
  }
}


lib/sunrise_sunset.dart
library sunrise_sunset;

export 'src/sunrise_sunset_data.dart';
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'src/sunrise_sunset_response.dart';

class SunriseSunset {
  /// Get results from the Sunrise Sunset API for the given [date], [latitude], and [longitude].
  static Future<SunriseSunsetResponse> getResults({
    DateTime date,
    double latitude,
    double longitude,
  }) async {
    try {
      Uri uri = Uri(
        scheme: 'https',
        host: 'api.sunrise-sunset.org',
        path: '/json',
        queryParameters: {
          'date':
              date != null ? "${date.year}-${date.month}-${date.day}" : 'today',
          'lat': latitude.toString(),
          'lng': longitude.toString(),
          'formatted': '0'
        },
      );
      print('Calling ${uri.toString()}');

      http.Response response;
      try {
        response = await http.get(uri);
      } on Exception catch (e) {
        print('Network Error: $e');
        throw e;
      }

      final jsonResponse = json.decode(response.body);
      return SunriseSunsetResponse.fromJSON(jsonResponse);
    } catch (err) {
      print('Problem with URI: $err');
    }
    return null;
  }
}


lib/src/service/options.dart
///Options we use as params to the web service

class Options {
  /// [Optional] Date for which we want to obtain the API data.
  /// If it's null then we use the current date
  final DateTime date;

  final double latitude;

  final double longitude;

  ///[true] if we want the response from the API very well formatted.
  final bool formatted;

  Options({
    DateTime date,
    double latitude,
    double longitude,
    this.formatted = false,
  })  : this.date = date ?? DateTime.now(),
        this.latitude = _checkLatitude(latitude),
        this.longitude = _checkLongitude(longitude),
        assert(formatted != null);

  Map<String, dynamic> toQueryParams() {
    Map<String, dynamic> params = {};
    params['lat'] = latitude;
    params['lon'] = longitude;
    //DateFormat : yyyy-MM-dd
    params['date'] = "${date.year}-${date.month}-${date.day}";
    params['formatted'] = formatted ? '1' : '0';
    return params;
  }

  static double _checkLongitude(double longitude) {
    if (longitude != null && longitude >= -180.0 && longitude <= 180.0)
      return longitude;
    throw ArgumentError("Invalid longitude value of $longitude.");
  }

  static double _checkLatitude(double latitude) {
    if (latitude != null && latitude >= -180.0 && latitude <= 180.0)
      return latitude;
    throw ArgumentError("Invalid latitude value of $latitude.");
  }
}


lib/src/service/data_service.dart
import 'dart:convert';
import '../../sunrise_sunset.dart';
import './options.dart';
import 'package:http/http.dart' as http;

abstract class DataRepository {
  Future<SunriseSunsetResponse<T>> serverRequest<T>(
    Options options,
  );
}

class DataService implements DataRepository {
  /// HTTP network requests protocols.
  Future<SunriseSunsetResponse<T>> serverRequest<T>(
    Options options,
  ) async {
    Map<String, dynamic> queryParams = options.toQueryParams();

    Uri uri = Uri(
      scheme: 'https',
      host: '/api.sunrise-sunset.org',
      path: '/json',
      queryParameters: queryParams,
    );

    print('Calling ${uri.toString()}');

    http.Response response;
    try {
      response = await http.get(uri);
    } on Exception catch (e) {
      print('Network Error: $e');
      throw e;
    }

    final jsonResponse = json.decode(response.body);
    return parseData(jsonResponse, options);
  }
}

SunriseSunsetResponse parseData<T>(jsonResponse, Options options) {
  final status = jsonResponse["status"];
  final results = jsonResponse["results"];
  String errorMessage = "";

  if (status == "OK") {
  } else if (status == "INVALID_REQUEST") {
    errorMessage = "lat or lng parameters are missing or invalid";
  } else if (status == "INVALID_DATE") {
    errorMessage = "date parameter is missing or invalid";
  } else {
    errorMessage =
        "the request could not be processed due to a server error. The request may succeed if you try again";
  }

  if (options.formatted) {
    return SunriseSunsetResponse<SunriseSunsetDataFormatted>(
      data: errorMessage.isEmpty
          ? SunriseSunsetDataFormatted.fromJSON(results)
          : null,
      error: errorMessage,
      success: errorMessage.isEmpty,
    );
  } else {
    return SunriseSunsetResponse<SunriseSunsetData>(
      data: errorMessage.isEmpty ? SunriseSunsetData.fromJSON(results) : null,
      error: errorMessage,
      success: errorMessage.isEmpty,
    );
  }
}

class SunriseSunsetResponse<T> {
  final T data;
  final bool success;
  final String error;

  SunriseSunsetResponse({
    this.data,
    this.success,
    this.error,
  });
}


lib/src/sunrise_sunset_response.dart
import './sunrise_sunset_data.dart';

class SunriseSunsetResponse {
  final SunriseSunsetData data;
  final bool success;
  final String error;

  SunriseSunsetResponse({
    this.data,
    this.success,
    this.error,
  });

  factory SunriseSunsetResponse.fromJSON(Map json) {
    final status = json["status"];
    final results = json["results"];
    String errorMessage = "";

    if (status == "OK") {
    } else if (status == "INVALID_REQUEST") {
      errorMessage = "lat or lng parameters are missing or invalid";
    } else if (status == "INVALID_DATE") {
      errorMessage = "date parameter is missing or invalid";
    } else {
      errorMessage =
          "the request could not be processed due to a server error. The request may succeed if you try again";
    }

    return SunriseSunsetResponse(
      data: errorMessage.isEmpty ? SunriseSunsetData.fromJSON(results) : null,
      error: errorMessage,
      success: errorMessage?.isEmpty,
    );
  }
}


lib/src/model/sunrise_sunset_data_formatted.dart
///This class wrap the response from API formatted
///
///
/// Sample data:
///
/// {
///   "results":
///   {
///     "sunrise":"7:27:02 AM",
///     "sunset":"5:05:55 PM",
///     "solar_noon":"12:16:28 PM",
///     "day_length":"9:38:53",
///     "civil_twilight_begin":"6:58:14 AM",
///     "civil_twilight_end":"5:34:43 PM",
///     "nautical_twilight_begin":"6:25:47 AM",
///     "nautical_twilight_end":"6:07:10 PM",
///     "astronomical_twilight_begin":"5:54:14 AM",
///     "astronomical_twilight_end":"6:38:43 PM"
///   },
///    "status":"OK"
/// }
///
///
///
///
class SunriseSunsetDataFormatted {
  final String sunrise;
  final String sunset;
  final String solarNoon;
  final String dayLength;
  final String civilTwilightBegin;
  final String civilTwilightEnd;
  final String nauticalTwilightBegin;
  final String nauticalTwilightEnd;
  final String astronomicalTwilightBegin;
  final String astronomicalTwilightEnd;

  SunriseSunsetDataFormatted._({
    this.sunrise,
    this.sunset,
    this.solarNoon,
    this.dayLength,
    this.civilTwilightBegin,
    this.civilTwilightEnd,
    this.nauticalTwilightBegin,
    this.nauticalTwilightEnd,
    this.astronomicalTwilightBegin,
    this.astronomicalTwilightEnd,
  });

  factory SunriseSunsetDataFormatted.fromJSON(Map json) {
    return SunriseSunsetDataFormatted._(
      sunrise: json["sunrise"],
      sunset: json["sunset"],
      solarNoon: json["solar_noon"],
      civilTwilightBegin: json["civil_twilight_begin"],
      civilTwilightEnd: json["civil_twilight_end"],
      nauticalTwilightBegin: json["nautical_twilight_begin"],
      nauticalTwilightEnd: json["nautical_twilight_end"],
      astronomicalTwilightBegin: json["astronomical_twilight_begin"],
      astronomicalTwilightEnd: json["astronomical_twilight_end"],
      dayLength: json["day_length"],
    );
  }
}


lib/src/model/sunrise_sunset_data.dart
///This class wrap the response from API without formatting
///
///
/// Sample data:
///
/// {
///   "results":
///   {
///     "sunrise":"2015-05-21T05:05:35+00:00",
///     "sunset":"2015-05-21T19:22:59+00:00",
///     "solar_noon":"2015-05-21T12:14:17+00:00",
///     "day_length":51444,
///     "civil_twilight_begin":"2015-05-21T04:36:17+00:00",
///     "civil_twilight_end":"2015-05-21T19:52:17+00:00",
///     "nautical_twilight_begin":"2015-05-21T04:00:13+00:00",
///     "nautical_twilight_end":"2015-05-21T20:28:21+00:00",
///     "astronomical_twilight_begin":"2015-05-21T03:20:49+00:00",
///     "astronomical_twilight_end":"2015-05-21T21:07:45+00:00"
///   },
///    "status":"OK"
/// }
///
///
///
///
class SunriseSunsetData {
  final DateTime sunrise;
  final DateTime sunset;
  final DateTime solarNoon;

  ///Expressed in seconds
  final int dayLength;
  final DateTime civilTwilightBegin;
  final DateTime civilTwilightEnd;
  final DateTime nauticalTwilightBegin;
  final DateTime nauticalTwilightEnd;
  final DateTime astronomicalTwilightBegin;
  final DateTime astronomicalTwilightEnd;

  SunriseSunsetData._({
    this.sunrise,
    this.sunset,
    this.solarNoon,
    this.dayLength,
    this.civilTwilightBegin,
    this.civilTwilightEnd,
    this.nauticalTwilightBegin,
    this.nauticalTwilightEnd,
    this.astronomicalTwilightBegin,
    this.astronomicalTwilightEnd,
  });

  factory SunriseSunsetData.fromJSON(Map json) {
    return SunriseSunsetData._(
      sunrise: DateTime.parse(
        json["sunrise"],
      ),
      sunset: DateTime.parse(
        json["sunset"],
      ),
      solarNoon: DateTime.parse(
        json["solar_noon"],
      ),
      civilTwilightBegin: DateTime.parse(
        json["civil_twilight_begin"],
      ),
      civilTwilightEnd: DateTime.parse(
        json["civil_twilight_end"],
      ),
      nauticalTwilightBegin: DateTime.parse(
        json["nautical_twilight_begin"],
      ),
      nauticalTwilightEnd: DateTime.parse(
        json["nautical_twilight_end"],
      ),
      astronomicalTwilightBegin: DateTime.parse(
        json["astronomical_twilight_begin"],
      ),
      astronomicalTwilightEnd: DateTime.parse(
        json["astronomical_twilight_end"],
      ),
      dayLength: json["day_length"],
    );
  }
}
